src/mjolnir/graphfilter.cc

```diff
@@ -219,7 +393,359 @@ void FilterTiles(GraphReader& reader,
+
+  LOG_INFO("Validating edges for aggregation");
+  // Iterate through all tiles in the local level
+  auto local_tiles = reader.GetTileSet(TileHierarchy::levels().back().level);
+  for (const auto& tile_id : local_tiles) {
+    // Get the graph tile. Read from this tile to create the new tile.
+    graph_tile_ptr tile = reader.GetGraphTile(tile_id);
+    assert(tile);
+
+    std::unordered_set<GraphId> processed_nodes;
+    std::unordered_set<uint64_t> no_agg_ways;
+    processed_nodes.reserve(tile->header()->nodecount());
+    no_agg_ways.reserve(tile->header()->directededgecount());
+
+    GraphId nodeid = GraphId(tile_id.tileid(), tile_id.level(), 0);
+    for (uint32_t i = 0; i < tile->header()->nodecount(); ++i, ++nodeid) {
+      const NodeInfo* nodeinfo = tile->node(i);
+      uint32_t idx = nodeinfo->edge_index();
+      for (uint32_t j = 0; j < nodeinfo->edge_count(); j++, idx++) {
+        const DirectedEdge* directededge = tile->directededge(idx);
+        if (processed_nodes.find(nodeid) == processed_nodes.end()) {
+          GraphId en = directededge->endnode();
+          std::list<PointLL> shape;
+          // check if we can aggregate the edges at this node.
+          ValidateData(reader, shape, en, processed_nodes, no_agg_ways, nodeid, tile, directededge);
+        }
+      }
+    }
+
+    if (reader.OverCommitted()) {
+      reader.Trim();
+    }
+
+    // Now loop again double checking the ways.
+    nodeid = GraphId(tile_id.tileid(), tile_id.level(), 0);
+    for (uint32_t i = 0; i < tile->header()->nodecount(); ++i, ++nodeid) {
+      const NodeInfo* nodeinfo = tile->node(i);
+      uint32_t idx = nodeinfo->edge_index();
+      for (uint32_t j = 0; j < nodeinfo->edge_count(); j++, idx++) {
+        const DirectedEdge* directededge = tile->directededge(idx);
+        if (no_agg_ways.find(tile->edgeinfo(directededge).wayid()) != no_agg_ways.end()) {
+          processed_nodes.insert(directededge->endnode());
+        }
+      }
+    }
+
+    if (reader.OverCommitted()) {
+      reader.Trim();
+    }
+
+    // Create a new tile builder
+    GraphTileBuilder tilebuilder(reader.tile_dir(), tile_id, false);
+    std::vector<NodeInfo> nodes;
+
+    // Copy edges (they do not change)
+    std::vector<DirectedEdge> directededges;
+    size_t n = tile->header()->directededgecount();
+    directededges.reserve(n);
+    const DirectedEdge* orig_edges = tile->directededge(0);
+    std::copy(orig_edges, orig_edges + n, std::back_inserter(directededges));
+
+    nodeid = GraphId(tile_id.tileid(), tile_id.level(), 0);
+    for (uint32_t i = 0; i < tile->header()->nodecount(); ++i, ++nodeid) {
+      NodeInfo nodeinfo = tilebuilder.node(i);
+      bool found = (processed_nodes.find(nodeid) != processed_nodes.end());
+
+      // We can not aggregate at this node.  Turn off the mode change(aggregation) bit
+      if (found) {
+        nodeinfo.set_mode_change(false);
```

Comment:
can we leave a comment on this bit in nodeinfo that we are repurosing it temporarily for this filtering use-case